Build static library for JavaScriptCore before linking framework

ddkilzer opened the pull request:
#### 3d732906995061dc7c611017fbc04e3dd344ede5
<pre>
Build static library for JavaScriptCore before linking framework
<a href="https://bugs.webkit.org/show_bug.cgi?id=263697">https://bugs.webkit.org/show_bug.cgi?id=263697</a>
&lt;<a href="https://rdar.apple.com/113641596">rdar://113641596</a>&gt;

Reviewed by Elliott Williams.

This change introduces a libJavaScriptCore.a target, which is then used
to link JavaScriptCore.framework.  By configuring the project to build
this way, there is no need to duplicate the list of sources between both
targets.  When JSC_USE_STATIC_LINKING=YES, the jsc binary is linked to
libJavaScriptCore.a instead of to JavaScriptCore.framework.

* Source/JavaScriptCore/API/JSWrapperMap.h:
* Source/JavaScriptCore/API/ObjCCallbackFunction.h:
- Fix header include paths to work with static library.
* Source/JavaScriptCore/Configurations/BaseTarget.xcconfig: Copy from Source/JavaScriptCore/Configurations/JSC.xcconfig.
- Extract common variables from JavaScriptCore.xcconfig that are shared
  with libJavaScriptCore.xcconfig.
* Source/JavaScriptCore/Configurations/Base.xcconfig:
- Extract linker variables from JavaScriptCore.xcconfig that are shared
  with JSC.xcconfig.
- Note that `-fobjc-link-runtime` replaces
  `-lobjc -framework CoreFoundation -framework Foundation`.
* Source/JavaScriptCore/Configurations/JSC.xcconfig:
(JSC_USE_STATIC_LINKING): Add.
- Set to the same boolean value as ENABLE_LIBFUZZER (which will default
  to NO), but make it a separate variable to allow this behavior to be
  overriden via xcodebuild.
- Configure OTHER_LDFLAGS to statically link libJavaScriptCore.a when
  JSC_USE_STATIC_LINKING=YES, else use dynamic linking (default) to
  JavaScriptCore.framework.
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
- Extract shared variables into BaseTarget.xcconfig.
- Move build related variables into libJavaScriptCore.xcconfig.
* Source/JavaScriptCore/Configurations/libJavaScriptCore.xcconfig: Add.
- Move build-related variables from JavaScriptCore.xcconfig.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
- Add new xcconfig files to project.
- Add libJavaScriptCore.a target and use libJavaScriptCore.xcconfig as
  its base.
- Move from JavaScriptCore framework target to libJavaScriptCore target:
  - Custom Build Rules.
  - Compile Sources.
  - Target Dependencies.
  - Copy Profiling Data build phase script.
- Add target dependency on libJavaScriptCore to JavaScriptCore.
- Add JavaScriptCoreFramework.cpp source file to JavaScriptCore
  framework target.
- Remove all items from &quot;Link Binary with Libraries&quot; for JavaScriptCore
  and jsc targets (except libedit.dylib on jsc) since this is handled
  via OTHER_LDFLAGS in xcconfig files now.
* Source/JavaScriptCore/JavaScriptCoreFramework.cpp: Add.
- Xcode requires at least one source file to compile to link a
  framework so create an empty one.

Canonical link: <a href="https://commits.webkit.org/270226@main">https://commits.webkit.org/270226@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/999fdfb9f9be041ddcb3e51cb5c6aaf13638e5c6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24906 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3449 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26159 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27022 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22858 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5136 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/888 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23159 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25150 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2480 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21490 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27602 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2177 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22426 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28568 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/21653 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22710 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22774 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26393 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/24158 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2132 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/429 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/31566 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3412 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6918 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2575 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/31548 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3175 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2473 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6596 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/empty.cpp:
```diff
@@ -0,0 +1,26 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// Used to build JavaScriptCore.framework and link libJavaScriptCore.a.
```
Hmm, can we call this JavaScriptCoreFramework.cpp instead of empty.cpp?

ddkilzer replied:
Sure.  That's easy to change.

emw-apple commented on Source/JavaScriptCore/Configurations/Base-JavaScriptCore.xcconfig:
There is an (admittedly confusing) naming convention in WebKit2 where we have a `BaseTarget.xcconfig` that is shared amongst some, but not all targets, so _maybe_ it would be better to name this BaseTarget.xcconfig as well? I'm not sure, but just wanted to mention it.

ddkilzer replied:
I don't really have a strong preference here, so I'm okay changing it.  How strong is your preference?

emw-apple replied:
Let's do it!

ddkilzer replied:
Done with next PR update.

emw-apple commented on Source/JavaScriptCore/Configurations/libJavaScriptCore.xcconfig:
```diff
@@ -0,0 +1,44 @@
+CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING = $(CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING_$(ENABLE_LLVM_PROFILE_GENERATION));
+CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING_ENABLE_LLVM_PROFILE_GENERATION = YES;
+
+SKIP_INSTALL = $(WK_NOT_$(FORCE_TOOL_INSTALL));
```
I wonder if this should be `NO`. When I experimented with creating a JavaScriptCore staticlib, it was multiple gigabytes in size. Assuming yours is similarly huge we should avoid ever installing it into the SDK.

ddkilzer replied:
Yes, it's 1.8 GB.  I'm fine changing this to `NO`.  I used the above since it was the convention for the targets.

```
$ du -sh Release/libJavaScriptCore.a 
1.8G	Release/libJavaScriptCore.a
```

ddkilzer replied:
Actually, I think you mean `SKIP_INSTALL` should be set to `YES`, not `NO`, to always be skipped?

emw-apple replied:
Yeah, you're right, make it `YES` :)

ddkilzer replied:
Done with next PR update.

emw-apple commented on Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
```diff
@@ -13549,6 +13592,38 @@
+			isa = XCBuildConfiguration;
+			baseConfigurationReference = 44F93DFD2AE71EBD00FFA37C /* libJavaScriptCore.xcconfig */;
+			buildSettings = {
+				PRODUCT_NAME = JavaScriptCore;
```
nit: You should be able to delete this PRODUCT_NAME setting and the four below (making `buildSettings` an empty dictionary). Xcode adds these by default but the definition is already in xcconfigs.

ddkilzer replied:
Done!

emw-apple commented on Source/JavaScriptCore/API/JSWrapperMap.h:
```diff
@@ -24,8 +24,8 @@
 #import <JavaScriptCore/JavaScriptCore.h>
-#import <JSValueInternal.h>
 #import <objc/objc-runtime.h>
+#import "JSValueInternal.h"
```
I think these need to be reordered to satisfy check-webkit-style.

ddkilzer replied:
Done!

emw-apple reviewed (approved):
r+ with a couple fixups. Looks great!

Commit:
Build static library for JavaScriptCore before linking framework
https://bugs.webkit.org/show_bug.cgi?id=263697
<rdar://113641596>

Reviewed by Elliott Williams.

This change introduces a libJavaScriptCore.a target, which is then used
to link JavaScriptCore.framework.  By configuring the project to build
this way, there is no need to duplicate the list of sources between both
targets.  When JSC_USE_STATIC_LINKING=YES, the jsc binary is linked to
libJavaScriptCore.a instead of to JavaScriptCore.framework.

* Source/JavaScriptCore/API/JSWrapperMap.h:
* Source/JavaScriptCore/API/ObjCCallbackFunction.h:
- Fix header include paths to work with static library.
* Source/JavaScriptCore/Configurations/BaseTarget.xcconfig: Copy from Source/JavaScriptCore/Configurations/JSC.xcconfig.
- Extract common variables from JavaScriptCore.xcconfig that are shared
  with libJavaScriptCore.xcconfig.
* Source/JavaScriptCore/Configurations/Base.xcconfig:
- Extract linker variables from JavaScriptCore.xcconfig that are shared
  with JSC.xcconfig.
- Note that `-fobjc-link-runtime` replaces
  `-lobjc -framework CoreFoundation -framework Foundation`.
* Source/JavaScriptCore/Configurations/JSC.xcconfig:
(JSC_USE_STATIC_LINKING): Add.
- Set to the same boolean value as ENABLE_LIBFUZZER (which will default
  to NO), but make it a separate variable to allow this behavior to be
  overriden via xcodebuild.
- Configure OTHER_LDFLAGS to statically link libJavaScriptCore.a when
  JSC_USE_STATIC_LINKING=YES, else use dynamic linking (default) to
  JavaScriptCore.framework.
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
- Extract shared variables into BaseTarget.xcconfig.
- Move build related variables into libJavaScriptCore.xcconfig.
* Source/JavaScriptCore/Configurations/libJavaScriptCore.xcconfig: Add.
- Move build-related variables from JavaScriptCore.xcconfig.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
- Add new xcconfig files to project.
- Add libJavaScriptCore.a target and use libJavaScriptCore.xcconfig as
  its base.
- Move from JavaScriptCore framework target to libJavaScriptCore target:
  - Custom Build Rules.
  - Compile Sources.
  - Target Dependencies.
  - Copy Profiling Data build phase script.
- Add target dependency on libJavaScriptCore to JavaScriptCore.
- Add JavaScriptCoreFramework.cpp source file to JavaScriptCore
  framework target.
- Remove all items from "Link Binary with Libraries" for JavaScriptCore
  and jsc targets (except libedit.dylib on jsc) since this is handled
  via OTHER_LDFLAGS in xcconfig files now.
* Source/JavaScriptCore/JavaScriptCoreFramework.cpp: Add.
- Xcode requires at least one source file to compile to link a
  framework so create an empty one.

Canonical link: https://commits.webkit.org/270226@main
